funnel/utils/markdown/mdit_plugins/abbr.py

```diff
@@ -0,0 +1,173 @@
+"""
+Markdown-it-py plugin to introduce <abbr> markup for defined abbreviations.
+
+Ported from javascript plugin markdown-it-abbr.
+"""
+
+import re
+import unicodedata
+
+from markdown_it import MarkdownIt
+from markdown_it.rules_block import StateBlock
+from markdown_it.rules_inline import StateInline
+from markdown_it.token import Token
+
+__all__ = ['abbr_plugin']
+
+ASTERISK_CHAR = 0x2A  # ASCII value for `*`
+OPEN_SQBR_CHAR = 0x5B  # ASCII value for `[`
+CLOSE_SQBR_CHAR = 0x5D  # ASCII value for `]`
+BACKSLASH_CHAR = 0x5C  # ASCII value for `\`
+COLON_CHAR = 0x3A  # ASCII value for `:`
+
+PADDING_CHARACTERS_RE = (
+    '['
+    + ''.join(
+        chr(x) for x in range(65536) if unicodedata.category(chr(x))[0] in ('P', 'Z')
+    )
+    .replace('\\', '\\\\')
+    .replace('[', '\\[')
+    .replace(']', '\\]')
+    + '\r\n$+<=>^`|~'
+    + ']'
+)
+
+
+def abbr_def(state: StateBlock, start_line: int, end_line: int, silent: bool):
+    pos = state.bMarks[start_line] + state.tShift[start_line]
+    maximum = state.eMarks[start_line]
+
+    if pos + 2 >= maximum:
+        return False
+
+    # Check if line is of the format `*[Abbreviation]:Full Form`
+    if state.srcCharCode[pos] != ASTERISK_CHAR:
+        return False
+    pos += 1
+    if state.srcCharCode[pos] != OPEN_SQBR_CHAR:
+        return False
+    pos += 1
+
+    label_end = -1
+    label_start = pos
+    while pos < maximum:
+        ch = state.srcCharCode[pos]
+        if ch == OPEN_SQBR_CHAR:  # pylint: disable=no-else-return
+            return False
+        elif ch == CLOSE_SQBR_CHAR:
+            label_end = pos
+            break
+        elif ch == BACKSLASH_CHAR:
+            pos += 1
+        pos += 1
+
+    if label_end < 0 or state.srcCharCode[label_end + 1] != COLON_CHAR:
+        return False
+
+    if silent:
+        return True
+
+    # Extract label and title and store it in state.env
+    label = state.src[label_start:label_end].replace('\\', '')
+    title = state.src[label_end + 2 : maximum].strip()
+    if len(label) == 0 or len(title) == 0:
+        return False
```

Comment:
Shouldn't this be a regex? It'll be far faster than char-by-char testing.

Reply:
Replacement regex:

`'^\s*\*\[(\w+)\]:\s*(.*)$'`

Reply:
I see `elif ch == BACKSLASH_CHAR:`, which means the following abbreviation is allowed:

`*[ab[b\]r]: Make ab[b]r into an abbreviation`

Reply:
IMO, this `ab[b]r` type example is not likely, and in the absence of an official spec for Markdown abbreviations, we can ignore these edge cases.

The `\w+` version allows an abbreviation to be a single word without spaces or hyphens or punctuation. This will not work for an abbreviation of the form `CAN-SPAM` because of the hyphen, so maybe we need the `[^\]]+` version allowing anything other than a `]`.

Or we could use `(?:[^\]|\])+` to allow the `\]` escape as well, giving an overall regex of:

`r'^\s*\*\[((?:[^\]|\])+)\]:\s*(.*)$'`

Reply:
Original implementation in PHP Markdown Extra: https://github.com/michelf/php-markdown/blob/932cbc7760983ce44f693fc3138f7ccebe685ccd/Michelf/MarkdownExtra.php#L1800-L1871